### PR TITLE
 add callback hook to extract gradients during training.

### PR DESCRIPTION
**Pull request by caidongqi**
This pull request introduces a new callback, `GradMonitor`, to the Composer library. The `GradMonitor` callback is designed to monitor gradients during training and store them in the state object. Additionally, the pull request includes updates to integrate this new callback with the existing training workflow and a new test file to demonstrate its usage.

Key changes include:

### New Callback Implementation:
* Added `GradMonitor` class in `composer/callbacks/grad_monitor.py` to extract and store gradients during training.

### Integration with Existing Codebase:
* Updated `composer/callbacks/__init__.py` to include `GradMonitor` in the list of available callbacks. [[1]](diffhunk://#diff-96a191bb7578802dcf75a72a7e4064e54478430e460068a64f51d8fcb3a82b32R28) [[2]](diffhunk://#diff-96a191bb7578802dcf75a72a7e4064e54478430e460068a64f51d8fcb3a82b32R50)
* Modified `composer/core/state.py` to include new attributes `grads`, `total_num_microbatches`, and `local_steps` to support the gradient monitoring functionality.
* Updated `composer/trainer/trainer.py` to set the `total_num_microbatches` in the training state.

### Testing and Demonstration:
* Added a new test file `test_demo.py` to demonstrate the usage of the `GradMonitor` callback with a toy convolutional neural network on the MNIST dataset.

### CODE REVIEW

File: composer/core/state.py [hunk 1]
```diff
@@ -503,7 +503,19 @@ def __init__(
         
         # Is the model of the fine-tuning type
         is_model_finetune: bool = False,
+
+        # average model grads of all batches in one global round
+        grads: Optional[list] = None,
```
Comment (Iacob-Alexandru-Andrei) [inline composer/core/state.py:508]: Typing has to be complete, using list here makes it impossible to use static typing properly later

File: composer/callbacks/grad_monitor.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        Args:
+            state (State): The state object.
+            device (torch.device, optional): The device to store the gradients. Defaults to CPU.
+        """
+        
+        group = list(state.model.parameters())
+        grad_list = []
+        for p in group:
+            if p.grad is not None:
+                grad_list.append(p.grad.detach().clone().to(device))
```
Comment (Iacob-Alexandru-Andrei) [inline composer/callbacks/grad_monitor.py:58]: Send to device first, then .detach() then .clone(), this guarantees that no GPU allocation is ever made, then that we separate it from the computational graph, then that we have a copy we can use.
